apps/dashboard/src/components/workflow-editor/workflow-canvas.tsx

```diff
@@ -0,0 +1,159 @@
+import { useCallback, useEffect, useLayoutEffect, useMemo, useRef } from 'react';
+import {
+  Background,
+  BackgroundVariant,
+  Controls,
+  Node,
+  ReactFlow,
+  ReactFlowProvider,
+  useReactFlow,
+} from '@xyflow/react';
+import type { StepResponseDto } from '@novu/shared';
+import '@xyflow/react/dist/style.css';
+import {
+  AddNode,
+  ChatNode,
+  CustomNode,
+  DelayNode,
+  DigestNode,
+  EmailNode,
+  InAppNode,
+  PushNode,
+  SmsNode,
+  TriggerNode,
+} from './nodes';
+import { AddNodeEdgeType, AddNodeEdge } from './edges';
+import { NODE_HEIGHT, NODE_WIDTH } from './base-node';
+
+const nodeTypes = {
+  trigger: TriggerNode,
+  email: EmailNode,
+  sms: SmsNode,
+  in_app: InAppNode,
+  push: PushNode,
+  chat: ChatNode,
+  delay: DelayNode,
+  digest: DigestNode,
+  custom: CustomNode,
+  add: AddNode,
+};
+
+const edgeTypes = {
+  addNode: AddNodeEdge,
+};
+
+const panOnDrag = [1, 2];
+
+// y distance = node height + space between nodes
+const Y_DISTANCE = NODE_HEIGHT + 50;
+
+const mapStepToNode = (
+  step: StepResponseDto,
+  previousPosition: { x: number; y: number }
+): Node<Record<string, unknown>, keyof typeof nodeTypes> => {
+  return {
+    id: step.stepUuid,
+    position: { x: previousPosition.x, y: previousPosition.y + Y_DISTANCE },
+    // TODO pass the step title and content
+    data: {},
+    type: step.type,
+  };
+};
+
+const WorkflowCanvasChild = ({ steps }: { steps: StepResponseDto[] }) => {
+  const reactFlowWrapper = useRef<HTMLDivElement>(null);
+  const reactFlowInstance = useReactFlow();
+
+  const nodes = useMemo(() => {
+    const triggerNode = { id: '0', position: { x: 0, y: 0 }, data: {}, type: 'trigger' };
```

Comment:
Nit-picky comment, you can define this node outside of the function since it's a constant.

Reply:
as far as I remember, react-flow mutates this object, so we would need to do the clone deep or smth, that's why I decided to keep it like this